### PR TITLE
feat: Add namespace id checks for messages

### PR DESCRIPTION
**Pull request by SweeXordious**
Closes: https://github.com/celestiaorg/celestia-app/issues/571 and https://github.com/celestiaorg/celestia-app/issues/596

### CODE REVIEW

**Comment by SweeXordious** [comment]
@evan-forbes Are these changes enough for prepare/process proposal? So that I can start adding tests for that.

File: app/split_shares.go [hunk 1]
```diff
@@ -60,11 +60,18 @@ func SplitShares(txConf client.TxConfig, squareSize uint64, data *core.Data) ([]
 		}
 
 		msg := authTx.GetMsgs()[0]
+
```
Comment (SweeXordious) [inline app/split_shares.go:63]:
```suggestion
```

File: app/test/process_proposal_test.go [hunk 2]
```diff
@@ -164,13 +165,76 @@ func TestMessageInclusionCheck(t *testing.T) {
...
+		name           string
+		namespace      namespace.ID
+		expectedResult abci.ResponseProcessProposal_Result
+	}
+
+	tests := []test{
+		{"transaction namespace id for message", consts.TxNamespaceID, abci.ResponseProcessProposal_REJECT},
+		{"evidence namespace id for message", consts.EvidenceNamespaceID, abci.ResponseProcessProposal_REJECT},
+		{"tail padding namespace id for message", consts.TailPaddingNamespaceID, abci.ResponseProcessProposal_REJECT},
+		//{"parity shares namespace id for message", consts.ParitySharesNamespaceID, abci.ResponseProcessProposal_REJECT},
```
Comment (SweeXordious) [inline app/test/process_proposal_test.go:184]: Issue:

File: app/test/process_proposal_test.go [hunk 3]
```diff
@@ -164,13 +165,76 @@ func TestMessageInclusionCheck(t *testing.T) {
...
+		// this case is commented because the shares are ordered. Then, when adding a message with namespace consts.ParitySharesNamespaceID,
+		// it panics as it is higher than the maximum allowed namespace ID.
+		// issue:
```
Comment (SweeXordious) [inline app/test/process_proposal_test.go:187]:
```suggestion
		// issue: https://github.com/celestiaorg/celestia-app/issues/596
```

File: app/test/process_proposal_test.go [hunk 4]
```diff
@@ -164,13 +165,76 @@ func TestMessageInclusionCheck(t *testing.T) {
...
+		res := testApp.ProcessProposal(input)
+		assert.Equal(t, tt.expectedResult, res.Result)
+	}
+}
+
 func genRandMsgPayForData(t *testing.T, signer *types.KeyringSigner, squareSize uint64) (*types.MsgPayForData, []byte) {
 	ns := make([]byte, consts.NamespaceSize)
 	_, err := rand.Read(ns)
 	require.NoError(t, err)
+	return genRandMsgPayForDataForNamespace(t, signer, squareSize, ns)
```
Comment (rootulp) [inline app/test/process_proposal_test.go:232]: 👍 great idea extracting this. The function name increases readability

File: app/test/prepare_proposal_test.go [hunk 5]
```diff
@@ -102,6 +104,39 @@ func TestPrepareProposal(t *testing.T) {
...
+		namespace        namespace.ID
+		expectedMessages int
+	}
+
+	tests := []test{
+		{"transaction namespace id for message", consts.TxNamespaceID, 0},
+		{"evidence namespace id for message", consts.EvidenceNamespaceID, 0},
+		{"tail padding namespace id for message", consts.TailPaddingNamespaceID, 0},
+		{"parity shares namespace id for message", consts.ParitySharesNamespaceID, 0},
+		{"namespace id 200 for message", namespace.ID{0, 0, 0, 0, 0, 0, 0, 200}, 0},
```
Comment (rootulp) [inline app/test/prepare_proposal_test.go:124]: [question] why does namespace id 200 have 0 messages? Is it because 200 is less than reserved namespace ID? If so, then maybe:

```suggestion
		{"reserved namespace id for message", namespace.ID{0, 0, 0, 0, 0, 0, 0, 200}, 0},
```

File: app/test/prepare_proposal_test.go [hunk 6]
```diff
@@ -102,6 +104,39 @@ func TestPrepareProposal(t *testing.T) {
...
+		{"correct namespace id for message", namespace.ID{3, 3, 2, 2, 2, 1, 1, 1}, 1},
```
Comment (rootulp) [inline app/test/prepare_proposal_test.go:125]: [out of curiosity] how was this namespace ID determined? Is it just an arbitrary namespace ID that is considered a valid namespace ID because it's the right length and doesn't use any reserved namespaces?

[nit]

```suggestion
		{"valid namespace id for message", namespace.ID{3, 3, 2, 2, 2, 1, 1, 1}, 1},
```

**Review by rootulp — APPROVED** [review]
LGTM. I proposed a few renames based on my understanding but all are optional. Defer to your judgement on what's clearer

File: app/test/process_proposal_test.go (hunk 2 above)
Comment (evan-forbes) [reply app/test/process_proposal_test.go:184]: can we still check for that without actually causing the panic? I _think_ its possible is the message is placed perfectly as the last share in the original square.

Comment (SweeXordious) [reply app/test/process_proposal_test.go:184]: yes, it is actually possible. But  would make the test a bit ugly I guess.
Currently, the panic happens when calculating the data availability root for the input block. If we skip that step, it will catch the wrong namespace id. 
To do that, we can add a parameter to the test struct called `calculateDARoot` of type `bool`. 
Something like this:

```golang
	type test struct {
		name            string
		namespace       namespace.ID
		calculateDARoot bool
		expectedResult  abci.ResponseProcessProposal_Result
	}

	tests := []test{
		{"transaction namespace id for message", consts.TxNamespaceID, true, abci.ResponseProcessProposal_REJECT},
		{"evidence namespace id for message", consts.EvidenceNamespaceID, true, abci.ResponseProcessProposal_REJECT},
		{"tail padding namespace id for message", consts.TailPaddingNamespaceID, true, abci.ResponseProcessProposal_REJECT},
		{"parity shares namespace id for message", consts.ParitySharesNamespaceID, false, abci.ResponseProcessProposal_REJECT},
		// this case is commented because the shares are ordered. Then, when adding a message with namespace consts.ParitySharesNamespaceID,
		// it panics as it is higher than the maximum allowed namespace ID.
		// issue:
		{"namespace id 200 for message", namespace.ID{0, 0, 0, 0, 0, 0, 0, 200}, true, abci.ResponseProcessProposal_REJECT},
		{"correct namespace id for message", namespace.ID{3, 3, 2, 2, 2, 1, 1, 1}, true, abci.ResponseProcessProposal_ACCEPT},
	}

	for _, tt := range tests {
		pfd, msg := genRandMsgPayForDataForNamespace(t, signer, 8, tt.namespace)
		input := abci.RequestProcessProposal{
			BlockData: &core.Data{
				Txs: [][]byte{
					buildTx(t, signer, encConf.TxConfig, pfd),
				},
				Messages: core.Messages{
					MessagesList: []*core.Message{
						{
							NamespaceId: pfd.GetMessageNamespaceId(),
							Data:        msg,
						},
					},
				},
				OriginalSquareSize: 8,
			},
		}
		if tt.calculateDARoot { <= here
			data, err := coretypes.DataFromProto(input.BlockData)
			require.NoError(t, err)

			shares, _, err := data.ComputeShares(input.BlockData.OriginalSquareSize)
			require.NoError(t, err)

			rawShares := shares.RawShares()

			require.NoError(t, err)
			eds, err := da.ExtendShares(input.BlockData.OriginalSquareSize, rawShares)
			require.NoError(t, err)
			dah := da.NewDataAvailabilityHeader(eds)
			input.Header.DataHash = dah.Hash()
		}
		res := testApp.ProcessProposal(input)
		assert.Equal(t, tt.expectedResult, res.Result)
	}
```

Comment (SweeXordious) [reply app/test/process_proposal_test.go:184]: if you like it this way, I can update the test

File: app/test/prepare_proposal_test.go (hunk 6 above)
Comment (SweeXordious) [reply app/test/prepare_proposal_test.go]: @rootulp yes 100%. We could also fuzz all reserved namespaces instead of providing only an example, etc. But, I guess it is too much for this. 
However, if it makes sense, we can create an issue to improve the test.

File: app/test/process_proposal_test.go (hunk 2 above)
Comment (evan-forbes) [reply app/test/process_proposal_test.go:184]: nah, as you mentioned in https://github.com/celestiaorg/celestia-app/issues/596#issuecomment-1205785617, this specific case isn't reachable

Comment (SweeXordious) [reply app/test/process_proposal_test.go:184]: But now that I think about it, I guess it's worth it. So that if someday, for some reason, we're no longer checking for parity shares namespace, we have this test to know exactly what's the issue?
I am okey with both, whichever you prefer.

File: x/payment/types/wirepayfordata.go [hunk 7]
```diff
@@ -139,6 +121,43 @@ func (msg *MsgWirePayForData) ValidateBasic() error {
...
+
+	// ensure that TxNamespaceID is not used
+	if bytes.Equal(ns, consts.TxNamespaceID) {
+		return ErrTxNamespace
+	}
+
+	// ensure that EvidenceNamespaceID is not used
+	if bytes.Equal(ns, consts.EvidenceNamespaceID) {
+		return ErrEvidenceNamespace
+	}
```
Comment (evan-forbes) [inline x/payment/types/wirepayfordata.go:156]: these checks are redundant to the one on line 134
```suggestion
```

File: x/payment/types/wirepayfordata.go [hunk 8]
```diff
@@ -84,12 +85,8 @@ func (msg *MsgWirePayForData) Route() string { return RouterKey }
...
 // interface.
 func (msg *MsgWirePayForData) ValidateBasic() error {
-	// ensure that the namespace id is of length == NamespaceIDSize
-	if nsLen := len(msg.GetMessageNameSpaceId()); nsLen != NamespaceIDSize {
-		return ErrInvalidNamespaceLen.Wrapf("got: %d want: %d",
-			nsLen,
-			NamespaceIDSize,
-		)
+	if err := ValidateMessageNamespaceID(msg.GetMessageNameSpaceId()); err != nil {
+		return err
```
Comment (evan-forbes) [inline x/payment/types/wirepayfordata.go:89]: :+1:

**Review by evan-forbes — APPROVED** [review]
found a redundant check, but pre-approving

File: app/test/process_proposal_test.go (hunk 2 above)
Comment (evan-forbes) [reply app/test/process_proposal_test.go:184]: > But now that I think about it, I guess it's worth it. So that if someday, for some reason, we're no longer checking for parity shares namespace, we have this test to know exactly what's the issue?

that's another great point. before reading your comment in #596, I was going to suggest that we split it off into its own test for this exact reason. Its a super explicit reminder of this edge case for the future generations to remember.

Comment (evan-forbes) [reply app/test/process_proposal_test.go:184]: then we can leave the original test nice and clleeeaaaannnn

Comment (SweeXordious) [reply app/test/process_proposal_test.go:184]: sounds awesome. Will do.

File: x/payment/types/wirepayfordata.go [hunk 9]
```diff
@@ -139,6 +121,34 @@ func (msg *MsgWirePayForData) ValidateBasic() error {
...
+	if bytes.Equal(ns, consts.ParitySharesNamespaceID) {
+		return ErrParitySharesNamespace
+	}
+
+	// ensure that TailPaddingNamespaceID is not used
+	if bytes.Equal(ns, consts.TailPaddingNamespaceID) {
+		return ErrTailPaddingNamespace
+	}
+
+
```
Comment (SweeXordious) [inline x/payment/types/wirepayfordata.go:148]:
```suggestion
```

File: app/test/prepare_proposal_test.go [hunk 10]
```diff
@@ -102,6 +104,39 @@ func TestPrepareProposal(t *testing.T) {
 	}
 }
 
+func TestPrepareMessagesWithReservedNamespaces(t *testing.T) {
```
Comment (evan-forbes) [inline app/test/prepare_proposal_test.go:107]: thanks for remembering this, I forgot lol

**Review by evan-forbes — APPROVED** [review]
:100: :1st_place_medal: :fire:

**Comment by evan-forbes** [comment]
just noticed that we are running `ValidateBasic` for the wirePFD twice during PrepareProposal, once for the message itself, and another time when we call `ValidateBasic` for the entire tx directly after as well. I like the explicitness of calling it directly, as we are relient upon the sdk always calling `ValidateBasic` for the msg for us, but :shrug:

**Comment by SweeXordious** [comment]
Just checked it, they are not the same `ValidateBasic`. One is the one we added, and the other one is this one:

https://github.com/celestiaorg/cosmos-sdk/blob/c20a3b1d3d68497f8e8fe8c1e20a12190246939f/types/tx/types.go#L34-L96

So, I guess it is good, no? or I am missing something.
